bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -788,89 +852,340 @@ private long readLastLogId(File f) {
         }
     }
 
-    /**
-     * Flushes all rotated log channels. After log channels are flushed,
-     * move leastUnflushedLogId ptr to current logId.
-     */
-    void checkpoint() throws IOException {
-        flushRotatedLogs();
+    interface EntryLogManager {
+
+        /*
+         * add entry to the corresponding entrylog and return the position of
+         * the entry in the entrylog
+         */
+        long addEntry(Long ledger, ByteBuf entry, boolean rollLog) throws IOException;
+
+        /*
+         * gets the active logChannel with the given entryLogId. null if it is
+         * not existing.
+         */
+        BufferedLogChannel getCurrentLogIfPresent(long entryLogId);
+
+        /*
+         * Returns eligible writable ledger dir for the creation next entrylog
+         */
+        File getDirForNextEntryLog(List<File> writableLedgerDirs);
+
+        /*
+         * Do the operations required for checkpoint.
+         */
+        void checkpoint() throws IOException;
+
+        /*
+         * roll entryLogs.
+         */
+        void rollLogs() throws IOException;
```

Comment:
after the change of `prepareMemtableFlush` and `commitMemtableFlush`, rollLogs seems to be only the behavior of single-log manager. I don't think you need `rollLogs` here, unless I missed something.

Reply:
rollLogs is not used in memtable flush related logic, but it is used in sortedledgerstorage.checkpoint logic. You made another comment related to rollLogs, I'm copying the same response..

i'm flexible here to make the change but

1) to begin with, from what i understand, this check (numBytesFlushed > 0) and rollLogs call is not needed. Since entries which come before checkpoint are already flushed from memtable (because of the way sortedledgerstorage.checkpoint) is called.
2) im not super convinced with moving from rollLogs method in interface to prepareCheckpoint, because this has meaning/use only for sortedledgerstorage and for single entrylog manager and having such a generic method name "prepareCheckpoint" seems to be a stretch.
3)in the method signature - "void prepareCheckpoint(Checkpoint checkpoint, long numBytesFlushedBetweenCheckpoint)" 'checkpoint' is not needed and we can remove it. parameter name 'numBytesFlushedBetweenCheckpoint' is not appropriate since it is just bytes added between previous flush and this checkpoint.

Reply:
okay that's find, it you want to leave `rollLogs` here, I am fine.